test/python/transpiler/test_sabre_swap.py

```diff
@@ -377,5 +384,879 @@ def test_rejects_too_few_qubits(self):
             pass_(qc)
 
 
+@ddt.ddt
+class TestSabreSwapControlFlow(QiskitTestCase):
+    """Tests for control flow in sabre swap."""
+
+    def test_shared_block(self):
+        """Test multiple control flow ops sharing the same block instance."""
+        inner = QuantumCircuit(2)
+        inner.cx(0, 1)
+
+        qreg = QuantumRegister(4, "q")
+        outer = QuantumCircuit(qreg, ClassicalRegister(1))
+        for pair in itertools.permutations(range(outer.num_qubits), 2):
+            outer.if_test((outer.cregs[0], 1), inner, pair, [])
+
+        coupling = CouplingMap.from_line(4)
+        cdag = SabreSwap(coupling, "lookahead", seed=82, trials=1).run(circuit_to_dag(outer))
+
+        check_map_pass = CheckMap(coupling)
+        check_map_pass.run(cdag)
+        self.assertTrue(check_map_pass.property_set["is_swap_mapped"])
+
+    def test_blocks_use_registers(self):
+        """Test that control flow ops using registers still use registers after routing."""
+        num_qubits = 2
+        qreg = QuantumRegister(num_qubits, "q")
+        cr1 = ClassicalRegister(1)
+        cr2 = ClassicalRegister(1)
+        qc = QuantumCircuit(qreg, cr1, cr2)
+
+        with qc.if_test((cr1, False)):
+            qc.cx(0, 1)
+            qc.measure(0, cr2)
+            with qc.if_test((cr2, True)):
+                qc.cx(0, 1)
```

Comment:
It's not so important, but `qc.measure(0, cr2)` and `if_test((cr2, True))` both look like type errors to me haha - the first is measure a bit into a register, and the second is comparing a register to a Boolean.  It happens to work because of a) the broadcast rules and b) `bool` subclassing `int` in Python, but it might be marginally cleaner to do
```suggestion
        cr1 = ClassicalRegister(1)
        cr2 = ClassicalRegister(1)
        qc = QuantumCircuit(qreg, cr1, cr2)

        with qc.if_test((cr1, False)):
            qc.cx(0, 1)
            qc.measure(0, cr2[0])
            with qc.if_test((cr2, 0)):
                qc.cx(0, 1)
```
just to avoid any possibility of us tightening the typing in the future.

Reply:
Heh, wouldn't breaking this code via tightening the typing be a backwards incompatible change? I'd hope we wouldn't do that, because I know we have a lot of instances of `qc.if_test((cr2, True)):` and probably even far more of `qc.x.c_if(cr2, True)` throughout the tests.